src/core/toga/widgets/table.py

```diff
@@ -25,127 +20,168 @@ def __repr__(self):
 
     @property
     def data(self):
-        '''
-        :returns: TableRow data
-        :rtype: ``data``
-        '''
+        """ TableRow data
+
+        Returns:
+            (``data``)
+        """
         return self._data
 
     @data.setter
     def data(self, data):
-        '''
-        TableRow data
-
-        :param data: Contains the row data
-        :type  data: ``dict``
-        '''
         self._data = data
-        if self.source.interface:
-            self.source.interface._impl.refresh()
 
     @property
     def icon(self):
-        '''
-        :returns: The image url of the row
-        :rtype: ``str``
-        '''
+        """ Icon on the row.
+        To set a image provide the path to the image as a ``str``.
+
+        Returns:
+            (str) The image url of the row as a ``str`.
+        """
         return self._icon
 
     @icon.setter
     def icon(self, path):
-        '''
-        Set an icon on the row
-
-        :param image_url: Url of the icon
-        :type  image_url: ``str``
-        '''
         if path is None:
             self._icon = None
         else:
             self._icon = Icon.load(path)
-            if self.source.interface:
-                self.source.interface._impl.refresh_row(self)
 
 
 class ListDataSource:
-    def __init__(self, data):
+    """ A data source that helps you to store and manage data in a row like fashion.
+
+    Args:
+        data (`list` of `tuple`): A list of tuples containing the data for every row.
+        refresh_function (`callable`): A function invoked on data change.
+    """
+
+    def __init__(self, data, on_refresh=None):
         self._data = self.create_rows(data)
-        self.interface = None
+        self._on_refresh = on_refresh if on_refresh else None
+        self._refresh_list = []
```

Comment:
What's the distinction between the "on_refresh" handler, and the "refresh_list" handlers? Aren't they the same thing? Couldn't we either (a) remove this on_refresh argument, or (b) make this a "listeners" argument which provides an initial list of listeners?